transformer_engine/pytorch/csrc/extensions/softmax.cu

```diff
@@ -245,3 +247,357 @@ at::Tensor scaled_aligned_causal_masked_softmax_backward(at::Tensor output_grad_
 
   return output_grads;
 }
+
+/***************************************************************************************************
+ * Support memory efficient cross entropy for Megatron-LM
+ **************************************************************************************************/
+template <typename dtype, int BlockSize>
+void __global__ CrossEntropyFwdSumExpKernel(float* sum_exp_logits_ptr,
+                                            dtype* vocab_parallel_logits_ptr, float* logits_max_ptr,
+                                            size_t n_dim) {
+  /***
+    1024 | 1
+    7 | 1016 | 2
+    6 | 1016 | 3
+
+    For example:
+    1024 | 1 -> [0,1023] [1024]
+    7 | 1016 | 2 -> [0, 6], [7,1022], [1023,1024]
```

Comment:
Could we have some text here in addition to the numbers? I do not quite understand what they represent.

Reply:
Sure, let me describe this process. @ptrendx 

  Because I will use vectorized reading in the upcoming calculations, reading in 8 bfloat16 at once.So I need to ensure that the read data is aligned with 16 bytes. In the cross-entropy computing scenario, ndim is usually a multiple of 8. 

  But in order to be compatible with future situations where the ndim may not be a multiple of 8, I used mathematical formulas to calculate the start and end offset addresses for each row that are aligned by 16 bytes. 
And for every row's misaligned situations, I will use a simple loop to handle them separately at the kernel's end.

  That is, ss our current threading model is a block responsible for computing one line, and for the case where the ndim is not an integer multiple of 8.Then we need to divide each row into three parts. The first and last part are not 16 byte aligned, and the middle part is 16 byte aligned. By this way, we can read the middle part according to vectorization and process the beginning and last part separately.
    

  For example, if ndim is 1025, because 1025 is not a multiple of 8, it is necessary to divide the 1025 data into "three parts". For the first block is responsible for calculating the first row, there are only two parts. The elements in positions [0-1023] are aligned with 16 bytes, but the last one needs to be processed separatel.For the second block is responsible for calculating the second row, there are only three parts. The elements in positions [7, 1022] are aligned with 16 bytes. but the [0, 6] and the [1023,1024] needs to be processed separatel.
    
  We can understand the calculation process of these indexes and values by combining them with the code.
  As shown in the following topology figure, 
    1024 | 1
    7 | 1016 | 2
    6 | 1016 | 3

  if the ndim = 1025, assuming we have three rows, the first row is `cur_vocab_parallel_logits_ptr_begin` = 0, `cur_vocab_parallel_logits_ptr_end` = 1025, `end_mol_num` = 1, `begin_mol_num` = 1024, `begin_offset` = 0, `end_offset` = 1023.
We define the 16 byte aligned region as [begin_offset, end_offset], for the first row, this region is [0, 1023]. That is, the first 1024 elements can be read by vectorized reading.

  Similarly, for the second row, `cur_vocab_parallel_logits_ptr_end` = 1025, `cur_vocab_parallel_logits_ptr_end` = 2050, `end_mol_num` = 2, `begin_mol_num` = 1023, `begin_offset` = 7, `end_offset` = 1022
We define the 16 byte aligned region as [begin_offset, end_offset], for the second row, this region is [7, 1022]. That is , the 1016 numbers in the middle. For the first 7 and last 2 numbers of the current row, they need to be processed separately because they are not aligned with 16 bytes.

  The calculation logic in the third row is also the same.

Reply:
Hmm, I did not yet read the actual note (will do in a second), but such explanations should be recorded somewhere possibly easier to find than a comment in a PR. At the very least we could move them to the PR description and in the code make a comment with a link to that PR?